add ability to copy meta/docinfo fields and aggregations into Logstash event

fbaligand opened the pull request:
Fixes #57 (copy response meta/docinfo fields to Logstash event)
Fixes #82 (copy response aggregations to Logstash event)

jsvd commented on lib/logstash/filters/elasticsearch.rb:
```diff
@@ -103,8 +103,14 @@ class LogStash::Filters::Elasticsearch < LogStash::Filters::Base
-  # Array of fields to copy from old event (found via elasticsearch) into new event
-  config :fields, :validate => :array, :default => {}
+  # Hash of fields to copy from old event (found via elasticsearch) into new event
+  config :fields, :validate => :hash, :default => {}
```
this represents a breaking change, the following config which worked before will not work anymore:

```
  filter {
    elasticsearch {
      fields => [["response", "code"]]
      #  ...
    }
  }
```

We can either revert this change or bump the plugin to 4.0.0, but since breaking changes are so costly when it comes to maintenance I'd suggest reverting. thoughts?

fbaligand replied:
Indeed, it is better to avoid breaking changes.
I did it to be consistent with actual plugin documentation (and the implementation code), which both uses this option as a hash (not as an array).
But ok, I would revert it to array.

fbaligand commented on docs/index.asciidoc:
```diff
@@ -133,6 +153,25 @@ filter plugins.
+  * Value type is <<hash,hash>>
+  * Default value is `{}`
+
+Hash of docinfo fields to copy from old event (found via elasticsearch) into new event
```
@jsvd 
For this option, there are 2 ways to see the point :
- use same semantics that elasticsearch input, which means also add `docinfo` and `docinfo_target` options.
- use same semantics that elasticsearch filter, which have only one `fields` option which maps es fields to ls fields.

Personally, I find better to have a consistent unique way to define mapping in elasticsearch filter between Elasticsearch and Logstash for options : fields/docinfo_fields/aggregation_fields.

=> Given these arguments, tell me which way you wish, and I would apply.

jsvd replied:
> use same semantics that elasticsearch filter, which have only one fields option which maps es fields to ls fields.

let's stick to this one, again for the reason of less breaking changes, we can plan to align them moving forward, specially when we start collapsing the repositories of {input/filter/output}-technology together into a single universal plugin for that technology (no more annoying mixins!)

fbaligand commented:
@jsvd 

OK, so I just pushed this :
- revert 'fields' option to array type
- given your answer, I changed nothing to `docinfo_fields`
- by the way, I removed documentation from code (plugin doc is now in docs/index.asciidoc)
- I also fixed doc example to be compatible with Logstash 5+

=> Is it all right for you ?

fbaligand commented:
@jsvd 
I just rebased my code to upstream/master.
Could you review and say if it is OK to merge ?

fbaligand commented:
Great ! 
Thanks @jsvd for this merge !

=> could you now release elasticsearch filter plugin ?

fbaligand commented:
@jsvd 
Please, could you release elasticsearch filter plugin ?

I need this enhancement for a project at my work.

jsvd commented:
[version v3.3.0 has been published](https://rubygems.org/gems/logstash-filter-elasticsearch/versions/3.3.0) to rubygems.org

fbaligand commented:
Great ! Thank you @jsvd !

webmat commented:
Hi @fbaligand, I was curious whether you've been using `aggregation_fields`, since this has been merged?

I just tried using it, and I got mixed results. All of the results of the aggregations do get added to the event, I find them pretty difficult to use.

Given a simple log event

```JSON
{
    "host": "matbook-pro.lan",
    "@timestamp": "2018-05-02T19:08:57.000Z",
    "type": "logstash",
    "path": "/Users/mat/work/elastic/logstash/local/input.log",
    "message": "ohai",
    "prog": "foo"
  }
```

And given the following filter-es config:

```
# logstash conf
elasticsearch {
  index => "logstash-*"
  query_template => "local/filter_es/count_logs.json"
  aggregation_fields => { "progs" => "test_agg" }
}

# local/filter_es/count_logs.json
{
  "size": 0,
  "query": {
    "query_string": {"query": "type:logstash"}
  },
  "aggs" : {
    "progs" : {
      "terms" : { "field" : "prog.keyword" }
    }
  }
}
```

The event I end up with does get a new `test_agg` attribute, but the `test_agg.buckets` attribute is an array of objects, which has limited ElasticSearch and Kibana support.

See also [this screenshot](https://www.dropbox.com/s/5pb5ilmvurbhj0d/Screenshot%202018-05-02%2015.11.21.png?dl=0) of Kibana complaining ;-)

So I'm curious if you've been using this. I think it would be very useful, but right now it seems like it would need a bit of tweaking, in order to be easy to use.

Note that I'm not asking you to fix this, I just want to see whether I'm missing something, before I queue this up as something I will work on.

And thanks for all your contributions to the plugins ecosystem :-)

webmat commented:
Excellent, thanks for the quick answer.

Yes it totally makes sense to go from what is stored as is, and keep processing it to make it look just right.
